extras/sht3x/sht3x.c

```diff
@@ -0,0 +1,443 @@
+        return NULL;
+    
+    // init sensor data structure
+    dev->bus  = bus;
+    dev->addr = addr;
+    dev->mode = single_shot;
+    dev->repeatability = high;
+    dev->meas_started = false;
+    dev->meas_start_time = 0;
+    
+    dev->active = true;
+
+    uint16_t status;
+
+    // soft-reset including status check leads to i2c problems.
+    // if (!sht3x_reset (dev))
+    if (!sht3x_get_status (dev, & status))
+    {
+	    error_dev ("could not reset or check the status of the sensor", __FUNCTION__, dev);
+        free(dev);
+        return NULL;       
+    }
+
+    // clear sensor status register
+    if (!sht3x_send_command(dev, SHT3x_CLEAR_STATUS_CMD))
+    {
+	    error_dev ("could not clear sensor status", __FUNCTION__, dev);
+        free(dev);
+        return NULL;       
+    }
+
+    // check the again the status after clear status command
+    if (!sht3x_get_status(dev, &status))
+    {
+	    error_dev ("could not get sensor status", __FUNCTION__, dev);
+        free(dev);
+        return NULL;       
+    }
+    
+    debug_dev ("sensor initialized", __FUNCTION__, dev);
+    return dev;
+}
+
+
+int32_t sht3x_start_measurement (sht3x_sensor_t* dev, sht3x_mode_t mode)
+{
+    if (!dev || !dev->active) return -1;
+    
+    dev->error_code = SHT3x_OK;
+    
+    // return remaining time when measurement is already running
+    if (dev->meas_started)
+    {
+        debug_dev ("measurement already started", __FUNCTION__, dev);
+        dev->error_code = SHT3x_MEAS_ALREADY_RUNNING;
+        return sht3x_is_measuring (dev);
+    }
+    
+    dev->mode = mode;
+    
+    // start measurement according to selected mode and return an duration estimate
+    if (sht3x_send_command(dev, SHT3x_MEASURE_CMD[dev->mode][dev->repeatability]))
+    {
+        dev->meas_start_time = sdk_system_get_time ();
+        dev->meas_started = true;
+        dev->meas_first = true;
+        return SHT3x_MEASURE_DURATION[dev->repeatability];   // in ms
+    }
+
+    dev->error_code |= SHT3x_SEND_MEAS_CMD_FAILED;
+    return -1;  // on error
+}
+
+
+// returns 0 when finished, -1 on error, remaining measurement time otherwise
+int32_t sht3x_is_measuring (sht3x_sensor_t* dev)
+{
+    if (!dev || !dev->active) return -1;
+
+    dev->error_code = SHT3x_OK;
+
+    if (!dev->meas_started)
+    {
+        error_dev ("measurement not started", __FUNCTION__, dev);
+        dev->error_code = SHT3x_MEAS_NOT_STARTED;
+        return -1;
+    }
+    
+    // computation is necessary in periodic mode only for first measurement
+    // in single shot mode every measurment is the first one
+    if (!dev->meas_first)
+        return 0;
+        
+    uint32_t start_time  = dev->meas_start_time ;     // in us
+	uint32_t system_time = sdk_system_get_time();   // in us
+	uint32_t elapsed_time;
+
+    if (system_time < start_time)
+        // in case of timer overflow
+        elapsed_time = (UINT32_MAX - start_time + system_time) / 1000;  // in ms
```

Comment:
This special handling for the timer overflow is not necessary. Simply subtracting two uint32 values will give the difference even if the timer has wrapped.

Reply:
Special handling removed.